velociraptor/catalogue/soap_catalogue.py

```diff
@@ -0,0 +1,266 @@
+                )
+                dynamically_register_properties(self.elements[-1])
+            elif isinstance(h5obj, h5py.Dataset):
+                self.elements.append(
+                    CatalogueDataset(self.file_name, f"{self.name}/{key}", handle)
+                )
+
+    def __str__(self):
+        return f"CatalogueGroup containing the following elements: {[el.name for el in self.elements]}"
+
+    def _register_properties(self):
+        self.properties = {}
+        for el in self.elements:
+            basename = el.name.split("/")[-1].lower()
+            # attribute names cannot start with a number
+            if basename[0].isnumeric():
+                basename = f"v{basename}"
+            if isinstance(el, CatalogueGroup):
+                setattr(self, basename, el)
+            elif isinstance(el, CatalogueDataset):
+                self.properties[basename] = (
+                    el,
+                    property(el.get_value, el.set_value, el.del_value),
+                )
+
+    def _register_extra_properties(self):
+        try:
+            stellar_velocity_dispersion_matrix = self.properties[
+                "stellarvelocitydispersionmatrix"
+            ][0]
+            self.elements.append(
+                VelocityDispersion(
+                    self.file_name,
+                    "stellarvelocitydispersion",
+                    [stellar_velocity_dispersion_matrix],
+                )
+            )
+            el = self.elements[-1]
+            self.properties["stellarvelocitydispersion"] = (
+                el,
+                property(el.get_value, el.set_value, el.del_value),
+            )
+        except KeyError:
+            pass
+
+
+def dynamically_register_properties(group: CatalogueGroup):
+    """
+    Absolute magic: trick an object into thinking it is of a different class
+    that has additional properties. Surprisingly, this works like a charm.
+    """
+
+    class_name = f"{group.__class__.__name__}{group.name.replace('/', '_')}"
+    props = {}
+    for name in group.properties:
+        props[name] = group.properties[name][1]
+    child_class = type(class_name, (group.__class__,), props)
+
+    group.__class__ = child_class
+
+
+class SOAPCatalogue(Catalogue):
+
+    file_name: str
+    root: CatalogueGroup
+    names_used: Set[str]
+
+    def __init__(self, file_name):
+        super().__init__("SOAP")
+        self.file_name = file_name
+        self.names_used = set()
+        self._register_quantities()
+
+    def print_fields(self):
+        print("SOAP catalogue fields used:")
+        for name in self.names_used:
+            print(f"  {name}")
+
+    def _register_quantities(self):
+        with h5py.File(self.file_name, "r") as handle:
+            self.root = CatalogueGroup(self.file_name, "", handle)
+            cosmology = handle["SWIFT/Cosmology"].attrs
+            self.units = SimpleNamespace()
```

Comment:
Yes, `catalogue.units` is required for compatibility with the old VR API. The pipeline uses things like `catalogue.units.box_length`. I am sure this can be changed to something more elegant in the future, but for now it serves its purpose.

Reply:
I mean for choosing SimpleNamespace instead of using a dataclass or a named tuple. Structured records like our units should use one of these instead of a SimpleNamespace.

Reply:
I don't know what dataclasses or named tuples are. Could you tell me how you would implement this?

Reply:
Well for the VR side we have a custom object that is the units: https://velociraptor-python.readthedocs.io/en/latest/_modules/velociraptor/units.html#VelociraptorUnits.

You could have SOAPUnits which is a class containing some of this information.

A dataclass is just a class but has less gubbins. It's basically a struct.

A namedtuple is even lighter weight, but it is immutable once created.